### PR TITLE
Allow collections to define their own action groups

### LINKED ISSUE #59864: module_defaults groups will not work with modules that only exist in collections
##### SUMMARY

Currently we maintain a [hard coded list](https://github.com/ansible/ansible/blob/80d8058f5c49c5d10eb3e527bd9d72052143a629/lib/ansible/config/module_defaults.yml) of module defaults. Modules distributed via collections have no way to use module defaults groups.

Furthermore, having a hard coded list does not scale well. New modules are often not added to the file and old modules are not removed from the file. Also, modules that are renamed or deprecated need to have the updates made to the module defaults file.

Ideally, this should come from the module itself via some mechanism, allowing module defaults to be user defined rather than dependent on a change to a config file that only exists in Ansible Core.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

`lib/ansible/config/module_defaults.yml`
##### ANSIBLE VERSION

```paste below
2.9
```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

```
ProductName:	Mac OS X
ProductVersion:	10.14.5
BuildVersion:	18F203
```

##### STEPS TO REPRODUCE

1. Create a module in a collection.
2. Try to use `module_defaults`

##### EXPECTED RESULTS

Module default values will work for modules distributed via a collection.

##### ACTUAL RESULTS

Module defaults do not work for the module that only exists within a collection.

### PR DESCRIPTION
**Pull request by s-hertel**
##### SUMMARY
Fixes #59864

Groupings should be in a collection's meta/action_groups.yml file in the format:
```
group_name:
  - module_name
```

##### ISSUE TYPE
- Feature Pull Request

### CODE REVIEW

File: lib/ansible/executor/module_common.py
Comment (s-hertel) [inline lib/ansible/executor/module_common.py]: may be better to just pass task at this point

Comment (bcoca) [reply lib/ansible/executor/module_common.py]: yep, but tempted to just wait and revamp the whole system, this shoudl just give us a transition to collections

File: lib/ansible/parsing/mod_args.py
Comment (bcoca) [inline lib/ansible/parsing/mod_args.py]: tempted to use dataloader .. but that might enable json and vaults

File: lib/ansible/parsing/mod_args.py
Comment (s-hertel) [inline lib/ansible/parsing/mod_args.py]: I should probably handle .yml | .yaml. Should I add json ?

File: lib/ansible/parsing/mod_args.py
Comment (nitzmahone) [inline lib/ansible/parsing/mod_args.py]: It looks like you're assuming the group lives in the same collection as the task action- it might not. The `module_defaults` declaration can be one of:
* `group/(FQCN of group name)` - eg, `group/community.aws.aws`
* `group/(unqualified group name)` - eg `group/aws`
* `(action/module FQCN)` - eg `community.aws.ec2`
* `(unqualified action/module name)` - eg `ec2`

The first case is straightforward- ask the named collection's metadata for the action group def, error if not found, party on. The second case should walk the collection list and ask each one for the group name, stopping at the first found definition (and later, once collection redirection is merged, consult the built-in unqualified redirect list before giving up). The third and fourth cases are basically just dynamic un-named groups of a single action/module.

This also points out to me that the lookup of the action groups and the storage of the defaults for the task probably belong in the playbook/DS hierarchy validation. That way, the metadata consultation is done in the controller (and the group metadata can be cached easily there), instead of repeatedly parsing and consulting the metadata in the workers.

**Review by nitzmahone — COMMENTED** [review]
Let's chat on this one- I think in order to do this in a performant manner, we need to move a little more stuff into the playbook/DS hierarchy validation rather than handling it post-fork in the worker. With a single shared config as it was originally, it made more sense where it is, but probably not anymore.

There are also a couple of things coming in the collection meta/redirection PR that you could leverage to get rid of a lot of code (eg we could just autoload the action_groups on initial collection load during playbook parse, and index them for snappy action lookups).

File: lib/ansible/parsing/mod_args.py
Comment (s-hertel) [reply lib/ansible/parsing/mod_args.py]: The second one feels kind of odd. I should expect that collections can use each other's action_groups? That seems like it would be a surprise to me as a user and would rather have things safely namespaced. Why not just inspect the collection of the task action and if it's not there, there is no group? The 1st's syntax to me seems pretty unfortunate but perhaps better than what I've done practically. - Edit: thinking about this some more, yes, I have to add the first syntax. Otherwise group names in different collections would get the same module_defaults arguments and would have the namespacing problem again. I'd question now whether group/(unqualified group name) should even be allowed, except that it would break current playbooks.

I did specifically avoided touching the action/module stuff since this PR was just to fix groupings, but I have looked into it and played with some possible solutions. FQCN works currently, except for ansible.builtin and ansible.legacy, which are just broken.

I asked bcoca about caching the results to improve performance and decided to hold off until I had something working decently. And yes, let's talk about the playbook/DS - the placement of this stuff feels a little awkward.

**Comment by geerlingguy** [comment]
Will there be any documentation on the format / purpose of the `action_groups` file? Is this something that's only really required for modules that are being migrated out from subdirs in ansible/ansible to collections, or is this something required (and/or optional?) for all collections?

**Comment by s-hertel** [comment]
@geerlingguy I will add documentation. The groups can be helpful for actions/modules that have a common set of options. For example, all of the AWS modules that take the same parameters for authentication are in the group `aws`, so instead of providing those for each task the default values for the modules in those could be set just once via module_defaults:
```
module_defaults:
  group/ansible.amazon.aws:
    aws_access_key: "{{ access_key }}"
    aws_secret_key: "{{ secret_key }}"
    aws_security_token: "{{ token }}"
    region: us-east-1
```
Here's the documentation for non-collection module_defaults groups: https://docs.ansible.com/ansible/latest/user_guide/playbooks_module_defaults.html#module-defaults-groups

**Comment by tremble** [comment]
> ```
> module_defaults:
>   group/ansible.amazon.aws:
> ```

I don't know how many examples there will be of this, but with the Community/Core split on the AWS side it would be nice **not** to _require_ people to use two separate group namespaces.  At the least we should allow the 'community' modules to use the 'core' group.

**Comment by s-hertel** [comment]
@tremble I agree. That example was really just to display the new syntax. `group/aws` will keep working for the things that have worked before. There was a comment in IRC that new things should only work with FQCN groups (which was not on my radar when I was working on this), but unless it's actively vetoed I plan on keeping `group/aws` working for new modules as well.

File: lib/ansible/executor/module_common.py
Comment (felixfontein) [inline lib/ansible/executor/module_common.py]: When I look at your above example:
```.yaml
   action_groups_redirection:
     another_group:
       redirect: exact_match
```
then `redirected_groups` would be a `dict`, and `redirected_group` will be `redirect`. If I use
```.yaml
   action_groups_redirection:
     another_group:
       - exact_match
```
instead, it works.

File: docs/docsite/rst/dev_guide/developing_collections.rst
Comment (felixfontein) [inline docs/docsite/rst/dev_guide/developing_collections.rst]: Here is the documentation I'm referring to.

Comment (s-hertel) [reply docs/docsite/rst/dev_guide/developing_collections.rst]: The documentation was correct, code was wrong

File: lib/ansible/executor/module_common.py
Comment (s-hertel) [reply lib/ansible/executor/module_common.py]: Fixed

**Review by felixfontein — APPROVED** [review]
Did some testing, seems to work fine!

**Comment by gundalow** [comment]
validation of `runtime.yml` done in https://github.com/ansible/ansible/pull/69742

File: lib/ansible/plugins/loader.py
Comment (nitzmahone) [inline lib/ansible/plugins/loader.py]: Vestigial?

File: lib/ansible/plugins/loader.py
Comment (nitzmahone) [inline lib/ansible/plugins/loader.py:598]: This might be a redirection bug, lemme think on if I should be handling this elsewhere

File: test/integration/targets/collections_action_groups/test_collections/ansible_collections/collection_namespace/collection_name/plugins/modules/example_module.py
Comment (nitzmahone) [inline test/integration/targets/collections_action_groups/test_collections/ansible_collections/collection_namespace/collection_name/plugins/modules/example_module.py]: probably better to use an echo module that returns everything it was given- also need an echo action

File: lib/ansible/parsing/mod_args.py
Comment (nitzmahone) [inline lib/ansible/parsing/mod_args.py]: This doesn't cover actions and breaks `skip_action_validation=True` in some cases. It's doing the right thing, but you can deconstruct the `has_plugin` calls from conditional above into `find_plugin_with_context` calls instead, then hang onto the result and use it here.

**Review by nitzmahone — CHANGES_REQUESTED** [review]
A few tweaks and a couple bugs (as we discussed), but looking good.

File: lib/ansible/parsing/mod_args.py
Comment (nitzmahone) [inline lib/ansible/parsing/mod_args.py]: check w/ @jimi-c on this one- it definitely works, but dunno if there's a better way or if there are risks to this approach I'm unaware of...

File: lib/ansible/plugins/loader.py
Comment (s-hertel) [reply lib/ansible/plugins/loader.py]: Thanks. Too many iterations of this thing :-)

File: test/integration/targets/collections_action_groups/test_collections/ansible_collections/collection_namespace/collection_name/plugins/modules/example_module.py
Comment (s-hertel) [reply test/integration/targets/collections_action_groups/test_collections/ansible_collections/collection_namespace/collection_name/plugins/modules/example_module.py]: Fixed. I added two action plugins, one with the same name as the module and one with a different name.

File: lib/ansible/parsing/mod_args.py
Comment (s-hertel) [reply lib/ansible/parsing/mod_args.py]: Improved this in 5e53832b683e20b202dededaccebffd37a45fb25

Thanks @bcoca and @jimi-c for the help!

**Comment by jillr** [comment]
Latest patchset tests successfully locally against community.vmware, community.aws, and amazon.aws. That aspect of this LGTM.

**Comment by felixfontein** [comment]
@s-hertel should collections continue to include `action_groups_redirection` in meta/runtime.yml for now, or should we remove this until the future of this feature is clear?

**Comment by felixfontein** [comment]
Also affects #69742 - should validation scream if `action_groups_redirection` is used, simply ignore it (and don't check its contents to make it possible to change it), or accept it only if it is in its current form?

**Comment by AlanCoding** [comment]
I tested this with the proposal for the `awx.awx` collection. See my notes in that PR https://github.com/ansible/awx/pull/7391

This is working as the documentation in it advertises 👍 

The functionality seems necessary for basic parity with Ansible 2.9. Collections have been emboldened to add content at a faster rate than pre-migration, so this will become even more important for users.

File: docs/docsite/rst/dev_guide/developing_collections.rst
Comment (AlanCoding) [inline docs/docsite/rst/dev_guide/developing_collections.rst:313]: Could you also document how to use this inside of a role? I would be highly interested in that.

**Comment by s-hertel** [comment]
@AlanCoding Thanks for taking a look at this. The portion of this PR responsible for backwards compatibility with 2.9 was moved to https://github.com/ansible/ansible/pull/69919 and merged for 2.10. I need to rebase since that was merged and I think it needs some more discussion.

**Comment by AlanCoding** [comment]
Yep, all on the same page. It looks like #70232 was following the pattern established in #69919. I don't want to take a position on that, because this is _old content_ but the _action group_ is new. `awx.awx` can wait in line for this fix to solve the problem generally.

It is a useful new feature to add, because every module in this collection needs authentication parameters. When running locally, they take env vars or a config file. Neither of these are particularly clean when running on a _remote_ host, which is common to do where the remote host is the actual server (so it makes API calls to localhost). Besides those options `block:` will work, but indents everything.

**Comment by AlanCoding** [comment]
Just a reminder that this would be awesome - but with heavy conflicts you may want to re-evaluate if it's worth keeping around.

**Comment by s-hertel** [comment]
@AlanCoding Thanks for the reminder. Still want to work on this, but I agree it will be quicker to start from scratch at this point. :-)

**Comment by felixfontein** [comment]
It's definitely awesome to be able to define action groups of the form `namespace.collection.action_group`.

I'm a bit wary about the feature to define short names in every collection, especially ones that are already in use by other collections. This would allow to create a `foo.bar` collection which has a useful module `left_pad` which coincidently grabs all module defaults, including cloud credentials of all kind, for various other modules and posts them to a random location on the web. (Well, I guess this can already happen now with all kind of plugins, but this would allow the same behavior for modules as well.)

**Pull request closed without merging**
